File: lib/loka/core/canonical.ex
defmodule Loka.Core.Canonical do
  @moduledoc """
  Strict JSON parser, canonical encoder and SHA-256 for the portable profile
  (`docs/spec/conformance/numeric-profile.md`).

  Values are `nil`, booleans, safe integers, UTF-8 binaries, lists and maps with ASCII
  binary keys. Both directions are linear in the input size.
  """
  @safe 9_007_199_254_740_991
  @max_depth 128

  @type value :: nil | boolean() | integer() | binary() | [value()] | %{binary() => value()}

  @doc "An integer in the profile's safe range, [-(2^53 - 1), 2^53 - 1]."
  defguard is_safe_integer(n) when is_integer(n) and abs(n) <= @safe

  @doc """
  Parses JSON text. Rejects duplicate or non-ASCII keys, fractions, exponents, NaN and
  Infinity, integers outside the safe range, lone surrogates, invalid UTF-8, nesting deeper
  than #{@max_depth} and trailing data. `-0` parses as `0`.
  """
  @spec decode(term()) :: {:ok, value()} | {:error, :invalid_json}
  def decode(text) when is_binary(text) do
    # String.valid? also rejects UTF-8-encoded surrogates.
    with true <- String.valid?(text), {v, rest} <- value(ws(text), 0), "" <- ws(rest) do
      {:ok, v}
    else
      _ -> {:error, :invalid_json}
    end
  catch
    :invalid -> {:error, :invalid_json}
  end

  def decode(_), do: {:error, :invalid_json}

  @doc "Canonical text: sorted keys, no whitespace. `:invalid_canonical` for a value outside the profile."
  @spec encode(term()) :: {:ok, binary()} | {:error, :invalid_canonical}
  def encode(value) do
    {:ok, :binary.list_to_bin([enc(value, 0)])}
  catch
    :invalid -> {:error, :invalid_canonical}
  end

  @doc "Lowercase hex SHA-256 of the canonical encoding."
  @spec hash(term()) :: {:ok, String.t()} | {:error, :invalid_canonical}
  def hash(value) do
    with {:ok, bytes} <- encode(value),
         do: {:ok, :crypto.hash(:sha256, bytes) |> Base.encode16(case: :lower)}
  end

  # ---- parser ----
  # `depth` counts the containers already open around the value being parsed.

  defp ws(<<c, rest::binary>>) when c in ~c" \t\n\r", do: ws(rest)
  defp ws(rest), do: rest

  defp value(<<c, _::binary>>, @max_depth) when c in ~c"{[", do: throw(:invalid)
  defp value(<<?{, rest::binary>>, depth), do: object(ws(rest), %{}, depth + 1)
  defp value(<<?[, rest::binary>>, depth), do: array(ws(rest), [], depth + 1)
  defp value(<<?", rest::binary>>, _), do: string(rest, rest, 0, [])
  defp value(<<"true", rest::binary>>, _), do: {true, rest}
  defp value(<<"false", rest::binary>>, _), do: {false, rest}
  defp value(<<"null", rest::binary>>, _), do: {nil, rest}
  defp value(<<?-, rest::binary>>, _), do: number(rest, -1)
  defp value(rest, _), do: number(rest, 1)

  defp object(<<?}, rest::binary>>, acc, _) when acc == %{}, do: {acc, rest}

  defp object(<<?", rest::binary>>, acc, depth) do
    {key, rest} = key(rest, acc)
    {v, rest} = value(colon(ws(rest)), depth)
    acc = Map.put(acc, key, v)

    case ws(rest) do
      <<?,, rest::binary>> -> object(ws(rest), acc, depth)
      <<?}, rest::binary>> -> {acc, rest}
      _ -> throw(:invalid)
    end
  end

  defp object(_, _, _), do: throw(:invalid)

  # Duplicates are detected after escape decoding.
  defp key(text, acc) do
    {key, rest} = string(text, text, 0, [])
    if Map.has_key?(acc, key) or not ascii?(key), do: throw(:invalid), else: {key, rest}
  end

  defp colon(<<?:, rest::binary>>), do: ws(rest)
  defp colon(_), do: throw(:invalid)

  defp array(<<?], rest::binary>>, [], _), do: {[], rest}

  defp array(text, acc, depth) do
    {v, rest} = value(text, depth)

    case ws(rest) do
      <<?,, rest::binary>> -> array(ws(rest), [v | acc], depth)
      <<?], rest::binary>> -> {Enum.reverse([v | acc]), rest}
      _ -> throw(:invalid)
    end
  end

  defp ascii?(key), do: not String.match?(key, ~r/[^\x00-\x7f]/)

  # `run` holds the pending literal bytes; they are copied in one piece at the next escape
  # or the closing quote.
  defp string(<<?", rest::binary>>, run, n, acc),
    do: {:binary.list_to_bin([acc | binary_part(run, 0, n)]), rest}

  defp string(<<?\\, rest::binary>>, run, n, acc) do
    {char, rest} = escape(rest)
    string(rest, rest, 0, [acc, binary_part(run, 0, n) | char])
  end

  defp string(<<c, rest::binary>>, run, n, acc) when c >= 0x20, do: string(rest, run, n + 1, acc)
  defp string(_, _, _, _), do: throw(:invalid)

  defp escape(<<?u, h::binary-4, rest::binary>>) do
    case {hex(h), rest} do
      {hi, <<"\\u", l::binary-4, rest::binary>>} when hi in 0xD800..0xDBFF ->
        lo = hex(l)
        if lo not in 0xDC00..0xDFFF, do: throw(:invalid)
        <<cp::utf16>> = <<hi::16, lo::16>>
        {<<cp::utf8>>, rest}

      {cp, _} when cp in 0xD800..0xDFFF ->
        throw(:invalid)

      {cp, _} ->
        {<<cp::utf8>>, rest}
    end
  end

  defp escape(<<c, rest::binary>>) when c in ~c'"\\/', do: {<<c>>, rest}
  defp escape(<<?b, rest::binary>>), do: {"\b", rest}
  defp escape(<<?f, rest::binary>>), do: {"\f", rest}
  defp escape(<<?n, rest::binary>>), do: {"\n", rest}
  defp escape(<<?r, rest::binary>>), do: {"\r", rest}
  defp escape(<<?t, rest::binary>>), do: {"\t", rest}
  defp escape(_), do: throw(:invalid)

  defp hex(digits) do
    if digits =~ ~r/\A[0-9a-fA-F]{4}\z/, do: String.to_integer(digits, 16), else: throw(:invalid)
  end

  # A fraction or exponent is left as trailing data, which every caller rejects. The
  # digit-count cap keeps String.to_integer off huge inputs.
  defp number(text, sign) do
    {digits, rest} =
      case text do
        <<?0, rest::binary>> -> {"0", rest}
        <<c, _::binary>> when c in ?1..?9 -> digits(text, text, 0)
        _ -> throw(:invalid)
      end

    n = if byte_size(digits) <= 16, do: sign * String.to_integer(digits), else: throw(:invalid)
    if is_safe_integer(n), do: {n, rest}, else: throw(:invalid)
  end

  defp digits(<<c, rest::binary>>, all, n) when c in ?0..?9, do: digits(rest, all, n + 1)
  defp digits(rest, all, n), do: {binary_part(all, 0, n), rest}

  # ---- encoder ----

  defp enc(nil, _), do: "null"
  defp enc(true, _), do: "true"
  defp enc(false, _), do: "false"
  defp enc(n, _) when is_safe_integer(n), do: Integer.to_string(n)

  defp enc(s, _) when is_binary(s) do
    if String.valid?(s), do: [?", esc(s, s, 0, []), ?"], else: throw(:invalid)
  end

  defp enc(v, @max_depth) when is_list(v) or is_map(v), do: throw(:invalid)

  defp enc(l, depth) when is_list(l),
    do: [?[, Enum.map_intersperse(l, ?,, &enc(&1, depth + 1)), ?]]

  defp enc(m, depth) when is_map(m) and not is_struct(m) do
    pairs =
      m
      |> Enum.sort()
      |> Enum.map_intersperse(?,, fn
        {k, v} when is_binary(k) ->
          if ascii?(k), do: [enc(k, depth), ?:, enc(v, depth + 1)], else: throw(:invalid)

        _ ->
          throw(:invalid)
      end)

    [?{, pairs, ?}]
  end

  defp enc(_, _), do: throw(:invalid)

  # Same run-copying as the parser: literal bytes leave in one piece.
  defp esc(<<>>, run, n, acc), do: [acc | binary_part(run, 0, n)]

  defp esc(<<c, rest::binary>>, run, n, acc) when c < 0x20 or c in ~c'"\\',
    do: esc(rest, rest, 0, [acc, binary_part(run, 0, n) | esc_char(c)])

  defp esc(<<_, rest::binary>>, run, n, acc), do: esc(rest, run, n + 1, acc)

  defp esc_char(?"), do: "\\\""
  defp esc_char(?\\), do: "\\\\"
  defp esc_char(?\b), do: "\\b"
  defp esc_char(?\t), do: "\\t"
  defp esc_char(?\n), do: "\\n"
  defp esc_char(?\f), do: "\\f"
  defp esc_char(?\r), do: "\\r"
  defp esc_char(c), do: ["\\u00" | Base.encode16(<<c>>, case: :lower)]
end

File: kernel/ts/src/canonical.ts
// Strict JSON parser, canonical encoder and hash for the portable profile
// (docs/spec/conformance/numeric-profile.md). Both directions are linear: literal runs are
// copied with one slice and output is joined once. Objects are null-prototype so keys such
// as "__proto__" stay ordinary data.
import { KernelError } from './error.ts';
import { sha256Hex, utf8 } from './sha256.ts';

export type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

export const SAFE = 9007199254740991;
/** Deepest container nesting the profile allows; one more is invalid_json / invalid_canonical. */
export const MAX_DEPTH = 128;

const invalid = (): never => {
  throw new KernelError('invalid_json');
};

const isSurrogate = (c: number): boolean => c >= 0xd800 && c <= 0xdfff;
const isHigh = (c: number): boolean => c >= 0xd800 && c <= 0xdbff;
const isLow = (c: number): boolean => c >= 0xdc00 && c <= 0xdfff;
const NON_ASCII = /[^\u0000-\u007f]/;

/**
 * Parses JSON text. Rejects duplicate or non-ASCII keys, fractions, exponents, NaN and
 * Infinity, integers outside the safe range, lone surrogates, nesting deeper than
 * MAX_DEPTH and trailing data. `-0` parses as 0. Throws KernelError('invalid_json').
 */
export function decode(text: string): Json {
  if (typeof text !== 'string') invalid();
  const p: Parser = { s: text, i: 0 };
  const v = value(p, 0);
  ws(p);
  if (p.i !== text.length) invalid();
  return v;
}

// Cursor over the text.
type Parser = { s: string; i: number };

function ws(p: Parser): void {
  for (let c = p.s[p.i]; c === ' ' || c === '\t' || c === '\n' || c === '\r'; c = p.s[++p.i]);
}

// `depth` counts the containers already open around the value.
function value(p: Parser, depth: number): Json {
  ws(p);
  const c = p.s[p.i];
  if ((c === '{' || c === '[') && depth === MAX_DEPTH) invalid();
  if (c === '{') return object(p, depth + 1);
  if (c === '[') return array(p, depth + 1);
  if (c === '"') return string(p);
  for (const [word, v] of LITERALS) {
    if (p.s.startsWith(word, p.i)) {
      p.i += word.length;
      return v;
    }
  }
  return number(p);
}

// Both containers: `{` or `[` at p.i, then members until `close`.
function members(p: Parser, close: string, member: () => void): void {
  p.i++;
  ws(p);
  if (p.s[p.i] === close) {
    p.i++;
    return;
  }
  for (;;) {
    member();
    ws(p);
    const c = p.s[p.i++];
    if (c === close) return;
    if (c !== ',') invalid();
  }
}

function object(p: Parser, depth: number): Json {
  const o: { [key: string]: Json } = Object.create(null);
  members(p, '}', () => {
    ws(p);
    if (p.s[p.i] !== '"') invalid();
    const k = string(p); // duplicates are detected after escape decoding
    if (NON_ASCII.test(k) || Object.hasOwn(o, k)) invalid();
    ws(p);
    if (p.s[p.i++] !== ':') invalid();
    o[k] = value(p, depth);
  });
  return o;
}

function array(p: Parser, depth: number): Json {
  const a: Json[] = [];
  members(p, ']', () => a.push(value(p, depth)));
  return a;
}

// Literal runs are copied with one slice at the next escape or the closing quote.
function string(p: Parser): string {
  const parts: string[] = [];
  let run = ++p.i;
  for (;;) {
    const c = p.s.charCodeAt(p.i);
    if (c === 0x22) break;
    if (!(c >= 0x20)) invalid(); // controls, and NaN past the end
    if (c === 0x5c) {
      parts.push(p.s.slice(run, p.i), escape(p));
      run = p.i;
    } else if (isSurrogate(c)) {
      if (!isHigh(c) || !isLow(p.s.charCodeAt(p.i + 1))) invalid();
      p.i += 2;
    } else p.i++;
  }
  parts.push(p.s.slice(run, p.i++));
  return parts.join('');
}

const ESCAPES: { [e: string]: string } = {
  '"': '"',
  '\\': '\\',
  '/': '/',
  b: '\b',
  f: '\f',
  n: '\n',
  r: '\r',
  t: '\t',
};

// `\` at p.i; returns the decoded character(s) and leaves p.i after the escape.
function escape(p: Parser): string {
  const e = p.s[p.i + 1];
  p.i += 2;
  if (e !== 'u') return Object.hasOwn(ESCAPES, e) ? ESCAPES[e] : invalid();
  const hi = hex4(p);
  if (isHigh(hi) && p.s.startsWith('\\u', p.i)) {
    p.i += 2;
    const lo = hex4(p);
    return isLow(lo) ? String.fromCharCode(hi, lo) : invalid();
  }
  return isSurrogate(hi) ? invalid() : String.fromCharCode(hi);
}

function hex4(p: Parser): number {
  const h = p.s.slice(p.i, (p.i += 4));
  return /^[0-9a-fA-F]{4}$/.test(h) ? parseInt(h, 16) : invalid();
}

// A fraction or exponent is left as trailing data, which every caller rejects. Above
// 2^53 - 1 the double may round, but only to a value that is still out of range.
function number(p: Parser): number {
  const start = p.i;
  if (p.s[p.i] === '-') p.i++;
  if (p.s[p.i] === '0') p.i++;
  else if (p.s[p.i] >= '1' && p.s[p.i] <= '9') while (p.s[p.i] >= '0' && p.s[p.i] <= '9') p.i++;
  else invalid();
  const n = Number(p.s.slice(start, p.i));
  return Math.abs(n) <= SAFE ? n + 0 : invalid(); // -0 becomes 0
}

const LITERALS: [string, Json][] = [
  ['true', true],
  ['false', false],
  ['null', null],
];

/** Canonical text: sorted keys, no whitespace. Throws KernelError('invalid_canonical') outside the profile. */
export function encode(value: Json): string {
  const out: string[] = [];
  write(value, out, 0);
  return out.join('');
}

/** Lowercase hex SHA-256 of the canonical encoding's UTF-8 bytes. */
export function hash(value: Json): string {
  return sha256Hex(utf8(encode(value)));
}

const notCanonical = (): never => {
  throw new KernelError('invalid_canonical');
};

function write(v: Json, out: string[], depth: number): void {
  if (v === null || v === true || v === false) out.push(String(v));
  else if (typeof v === 'number') {
    if (!Number.isInteger(v) || Math.abs(v) > SAFE) notCanonical();
    out.push(String(v + 0));
  } else if (typeof v === 'string') quote(v, out);
  else if (typeof v !== 'object' || depth === MAX_DEPTH) notCanonical();
  else if (Array.isArray(v)) {
    out.push('[');
    for (let n = 0; n < v.length; n++) {
      if (n) out.push(',');
      write(v[n], out, depth + 1);
    }
    out.push(']');
  } else {
    // ASCII keys only, so the default UTF-16 sort is ordinal order.
    const keys = Object.keys(v).sort();
    out.push('{');
    for (let n = 0; n < keys.length; n++) {
      if (NON_ASCII.test(keys[n])) notCanonical();
      if (n) out.push(',');
      quote(keys[n], out);
      out.push(':');
      write(v[keys[n]], out, depth + 1);
    }
    out.push('}');
  }
}

const SHORT: { [c: number]: string } = {
  0x22: '\\"',
  0x5c: '\\\\',
  0x08: '\\b',
  0x09: '\\t',
  0x0a: '\\n',
  0x0c: '\\f',
  0x0d: '\\r',
};

function quote(s: string, out: string[]): void {
  out.push('"');
  let run = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 0x20 || c === 0x22 || c === 0x5c) {
      out.push(s.slice(run, i), SHORT[c] ?? '\\u00' + c.toString(16).padStart(2, '0'));
      run = i + 1;
    } else if (isSurrogate(c)) {
      if (!isHigh(c) || !isLow(s.charCodeAt(i + 1))) notCanonical();
      i++;
    }
  }
  out.push(s.slice(run), '"');
}
